File: packages/core/src/summarize/summarize.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { summarizeRepo, buildSummarizePrompt } from './summarize.js';
import { getDb, closeDb } from '../db/index.js';
import { initDb } from '../db/init.js';

// Mock dependencies
vi.mock('../config/config.js', () => ({
  loadConfig: vi.fn(),
  isConfigured: vi.fn(),
}));

vi.mock('../gemini/llm.js', () => ({
  streamLLM: vi.fn(),
  isLLMError: vi.fn((result) => 'error' in result && !('text' in result)),
}));

// Import mocked modules
import { loadConfig } from '../config/config.js';
import { streamLLM } from '../gemini/llm.js';

const mockedLoadConfig = vi.mocked(loadConfig);
const mockedStreamLLM = vi.mocked(streamLLM);

describe('summarize/summarize', () => {
  const DB_PATH = ':memory:';

  beforeEach(() => {
    vi.clearAllMocks();

    // Default mocks
    mockedLoadConfig.mockReturnValue({
      githubPat: 'test-pat',
      geminiKey: 'test-gemini-key',
      dbPath: DB_PATH,
    });

    // Initialize DB
    initDb(DB_PATH);
    const db = getDb(DB_PATH);
    // Seed data
    db.prepare(`
      INSERT INTO repos (github_id, owner, name, full_name, url, pushed_at, synced_at)
      VALUES (1, 'owner', 'repo', 'owner/repo', 'https://github.com/owner/repo', datetime('now'), datetime('now'))
    `).run();

    db.prepare(`
      INSERT INTO chunks (repo_id, chunk_type, chunk_index, content)
      VALUES 
      (1, 'readme', 0, 'README content'),
      (1, 'file_tree', 0, 'src/index.ts'),
      (1, 'file_tree', 1, 'package.json')
    `).run();
  });

  afterEach(() => {
    closeDb();
    vi.clearAllMocks();
  });

  describe('summarizeRepo', () => {
    it('returns a populated SummarizeResult with summary and stats on success', async () => {
      mockedStreamLLM.mockResolvedValue({
        text: '## Overview\nSummary.\n## Tech Stack\nStack.\n## Use Cases\nCases.',
        inputTokens: 100,
        outputTokens: 50,
        durationMs: 500,
      });

      const result = await summarizeRepo({ repo: 'owner/repo' });

      expect(result.summary).toBe('## Overview\nSummary.\n## Tech Stack\nStack.\n## Use Cases\nCases.');
      expect(result.repo).toBe('owner/repo');
      expect(result.chunksUsed).toBe(3); // readme + 2 file_tree
      expect(result.inputTokens).toBe(100);
      expect(result.outputTokens).toBe(50);
      expect(result.durationMs).toBeGreaterThanOrEqual(0);
    });

    it('calls onChunk for each streamed chunk during summarizeRepo', async () => {
      mockedStreamLLM.mockImplementation(async (_apiKey, _prompt, _systemPrompt, onChunk) => {
        if (onChunk) {
          onChunk('Chunk 1');
          onChunk('Chunk 2');
        }
        return {
          text: 'Chunk 1Chunk 2',
          inputTokens: 10,
          outputTokens: 10,
          durationMs: 100,
        };
      });

      const chunks: string[] = [];
      await summarizeRepo({ repo: 'owner/repo' }, (text) => chunks.push(text));

      expect(chunks).toHaveLength(2);
      expect(chunks[0]).toBe('Chunk 1');
      expect(chunks[1]).toBe('Chunk 2');
    });

    it('fetches all chunks for the specified repo and includes them in context', async () => {
      mockedStreamLLM.mockResolvedValue({
        text: 'Summary',
        inputTokens: 10,
        outputTokens: 10,
        durationMs: 100,
      });

      await summarizeRepo({ repo: 'owner/repo' });

      expect(mockedStreamLLM).toHaveBeenCalled();
      const [, prompt] = mockedStreamLLM.mock.calls[0];
      
      expect(prompt).toContain('Repository: owner/repo');
      expect(prompt).toContain('--- readme ---');
      expect(prompt).toContain('README content');
      expect(prompt).toContain('--- file_tree ---');
      expect(prompt).toContain('src/index.ts');
      expect(prompt).toContain('package.json');
    });

    it('repo name matching is case-insensitive', async () => {
      mockedStreamLLM.mockResolvedValue({
        text: 'Summary',
        inputTokens: 10,
        outputTokens: 10,
        durationMs: 100,
      });

      const result = await summarizeRepo({ repo: 'Owner/Repo' });

      expect(result.chunksUsed).toBe(3);
      expect(result.repo).toBe('Owner/Repo'); // Returns the input repo name
    });

    it('returns safe result with no-data message when repo has no chunks', async () => {
      // Create a repo with no chunks
      const db = getDb(DB_PATH);
      db.prepare(`
        INSERT INTO repos (github_id, owner, name, full_name, url, pushed_at, synced_at)
        VALUES (2, 'empty', 'repo', 'empty/repo', 'https://github.com/empty/repo', datetime('now'), datetime('now'))
      `).run();

      const result = await summarizeRepo({ repo: 'empty/repo' });

      expect(result.summary).toContain('No data found for this repository');
      expect(result.chunksUsed).toBe(0);
      expect(mockedStreamLLM).not.toHaveBeenCalled();
    });
    
    it('returns safe result with no-data message when repo does not exist', async () => {
      const result = await summarizeRepo({ repo: 'nonexistent/repo' });

      expect(result.summary).toContain('No data found for this repository');
      expect(result.chunksUsed).toBe(0);
      expect(mockedStreamLLM).not.toHaveBeenCalled();
    });

    it('returns safe result with error message when streamLLM returns LLMError', async () => {
      mockedStreamLLM.mockResolvedValue({
        error: 'API error',
        durationMs: 100,
      });

      const result = await summarizeRepo({ repo: 'owner/repo' });

      expect(result.summary).toContain('Error generating summary: API error');
      expect(result.chunksUsed).toBe(3);
    });
    
    it('returns safe result with error message when not configured', async () => {
       mockedLoadConfig.mockReturnValue({
          githubPat: 'test-pat',
          geminiKey: null,
          dbPath: DB_PATH,
       });
       
       const result = await summarizeRepo({ repo: 'owner/repo' });
       
       expect(result.summary).toContain('Error: Gemini API key is not configured');
    });
  });

  describe('buildSummarizePrompt', () => {
    it('includes all chunk contents formatted with headers', () => {
      const chunks = [
        { chunk_type: 'readme', content: 'README' },
        { chunk_type: 'file_tree', content: 'TREE' }
      ];
      const prompt = buildSummarizePrompt('owner/repo', chunks);
      
      expect(prompt).toContain('Repository: owner/repo');
      expect(prompt).toContain('--- readme ---');
      expect(prompt).toContain('README');
      expect(prompt).toContain('--- file_tree ---');
      expect(prompt).toContain('TREE');
    });
  });
});

File: internal/summarize/summarize_test.go
package summarize

import (
	"context"
	"database/sql"
	"path/filepath"
	"testing"

	"github.com/hackastak/repog/internal/db"
	"github.com/hackastak/repog/internal/provider"
)

func TestBuildSummarizePrompt(t *testing.T) {
	chunks := []chunkRecord{
		{ChunkType: "metadata", Content: `{"full_name": "user/repo"}`},
		{ChunkType: "readme", Content: "# My Repo\nThis is a test repo."},
	}

	prompt := buildSummarizePrompt("user/repo", chunks)

	if !containsString(prompt, "user/repo") {
		t.Error("Prompt should contain repo name")
	}

	if !containsString(prompt, "metadata") {
		t.Error("Prompt should contain chunk type")
	}

	if !containsString(prompt, "# My Repo") {
		t.Error("Prompt should contain chunk content")
	}
}

func containsString(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func insertTestRepoWithChunks(t *testing.T, database *sql.DB, fullName string, chunks []chunkRecord) int64 {
	t.Helper()

	result, err := database.Exec(`
		INSERT INTO repos (github_id, full_name, owner, name, description, stars, language, topics, synced_at)
		VALUES (?, ?, 'owner', 'name', 'desc', 100, 'Go', '[]', datetime('now'))
	`, 12345, fullName)
	if err != nil {
		t.Fatalf("Failed to insert repo: %v", err)
	}

	repoID, err := result.LastInsertId()
	if err != nil {
		t.Fatalf("Failed to get repo ID: %v", err)
	}

	for _, chunk := range chunks {
		_, err := database.Exec(`
			INSERT INTO chunks (repo_id, chunk_type, content)
			VALUES (?, ?, ?)
		`, repoID, chunk.ChunkType, chunk.Content)
		if err != nil {
			t.Fatalf("Failed to insert chunk: %v", err)
		}
	}

	return repoID
}

func TestSummarizeRepo_ReturnsPopulatedResult(t *testing.T) {
	// Create mock LLM provider
	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			chunks := []string{
				"## Overview\nThis is a test repo.",
				"\n\n## Tech Stack\nGo language.",
				"\n\n## Use Cases\nTesting.",
			}
			fullText := ""
			for _, chunk := range chunks {
				fullText += chunk
				if onChunk != nil {
					onChunk(chunk)
				}
			}
			return &provider.LLMResult{Text: fullText, InputTokens: 100, OutputTokens: 50}, nil
		},
	}

	// Set up test database
	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Insert test data
	insertTestRepoWithChunks(t, database, "user/testrepo", []chunkRecord{
		{ChunkType: "metadata", Content: `{"full_name": "user/testrepo"}`},
		{ChunkType: "readme", Content: "# Test Repo\nA test repository."},
	})

	// Track chunks received
	var receivedChunks []string
	onChunk := func(chunk string) {
		receivedChunks = append(receivedChunks, chunk)
	}

	result, err := SummarizeRepo(context.Background(), SummarizeOptions{
		Repo:        "user/testrepo",
		DB:          database,
		LLMProvider: mockLLM,
	}, onChunk)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.Repo != "user/testrepo" {
		t.Errorf("Expected repo 'user/testrepo', got '%s'", result.Repo)
	}

	if result.ChunksUsed != 2 {
		t.Errorf("Expected 2 chunks used, got %d", result.ChunksUsed)
	}

	if !containsString(result.Summary, "## Overview") {
		t.Errorf("Expected summary to contain '## Overview', got '%s'", result.Summary)
	}

	if len(receivedChunks) == 0 {
		t.Error("Expected onChunk to be called")
	}

	if result.DurationMs < 0 {
		t.Errorf("Expected non-negative duration, got %d", result.DurationMs)
	}
}

func TestSummarizeRepo_CaseInsensitiveMatch(t *testing.T) {
	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			if onChunk != nil {
				onChunk("Summary")
			}
			return &provider.LLMResult{Text: "Summary", InputTokens: 10, OutputTokens: 5}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Insert with lowercase
	insertTestRepoWithChunks(t, database, "user/myrepo", []chunkRecord{
		{ChunkType: "readme", Content: "# My Repo"},
	})

	// Query with different case
	result, err := SummarizeRepo(context.Background(), SummarizeOptions{
		Repo:        "USER/MYREPO",
		DB:          database,
		LLMProvider: mockLLM,
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.ChunksUsed != 1 {
		t.Errorf("Expected 1 chunk used (case-insensitive match), got %d", result.ChunksUsed)
	}
}

func TestSummarizeRepo_ReturnsNoDataMessageWhenNoChunks(t *testing.T) {
	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	var receivedChunk string
	onChunk := func(chunk string) {
		receivedChunk = chunk
	}

	// LLM provider should not be called when no chunks exist
	mockLLM := provider.NewMockLLMProvider()

	result, err := SummarizeRepo(context.Background(), SummarizeOptions{
		Repo:        "nonexistent/repo",
		DB:          database,
		LLMProvider: mockLLM,
	}, onChunk)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	expectedMsg := "No data found for this repository. Try running `repog sync` and `repog embed` first."
	if result.Summary != expectedMsg {
		t.Errorf("Expected no data message, got '%s'", result.Summary)
	}

	if receivedChunk != expectedMsg {
		t.Errorf("Expected onChunk to receive no data message, got '%s'", receivedChunk)
	}

	if result.ChunksUsed != 0 {
		t.Errorf("Expected 0 chunks used, got %d", result.ChunksUsed)
	}
}

func TestSummarizeRepo_HandlesLLMError(t *testing.T) {
	// Create mock LLM provider that returns an error
	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, _ func(string)) (*provider.LLMResult, *provider.LLMError) {
			return nil, &provider.LLMError{Message: "Internal Server Error", StatusCode: 500}
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	insertTestRepoWithChunks(t, database, "user/repo", []chunkRecord{
		{ChunkType: "readme", Content: "# Repo"},
	})

	result, err := SummarizeRepo(context.Background(), SummarizeOptions{
		Repo:        "user/repo",
		DB:          database,
		LLMProvider: mockLLM,
	}, nil)

	// Should not return Go error, but Summary should contain error message
	if err != nil {
		t.Fatalf("Expected no Go error, got %v", err)
	}

	if !containsString(result.Summary, "Error generating summary") {
		t.Errorf("Expected error message in summary, got '%s'", result.Summary)
	}
}

func TestSummarizeRepo_NilOnChunkCallback(t *testing.T) {
	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			if onChunk != nil {
				onChunk("Summary")
			}
			return &provider.LLMResult{Text: "Summary", InputTokens: 10, OutputTokens: 5}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	insertTestRepoWithChunks(t, database, "user/repo", []chunkRecord{
		{ChunkType: "readme", Content: "# Repo"},
	})

	// Should not panic with nil callback
	result, err := SummarizeRepo(context.Background(), SummarizeOptions{
		Repo:        "user/repo",
		DB:          database,
		LLMProvider: mockLLM,
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.Summary != "Summary" {
		t.Errorf("Expected 'Summary', got '%s'", result.Summary)
	}
}
